### ISSUE TITLE
[feature] Checksum when downloading file

### ISSUE THREAD

**Issue by pquerna** [curl/curl#1399]
### I did this

I have a bunch of shell scripts and `Dockerfile`s.  They download files with curl.

Right now I have adhoc and inconsistent checksums of these downloaded files.

Rough example of this pattern:

```
export DLURL= "http://..."
export SOURCEHASH="c6596eb7be8581c18be736c846fb9173b69eccf6ef94c5135893ec56bd92ba08"
export LOCALPATH = "/tmp/example.tar.gz"

curl -s -o "${LOCALPATH}" "${DLURL}"
GOTHASH=$(openssl dgst -sha256 "${LOCALPATH}" | sed 's/^.* //')

if [ "${GOTHASH}" != "${SOURCEHASH}" ]; then
    echo "error: calculated hash ${GOTHASH} != provided hash: ${SOURCEHASH} for ${LOCALPATH}"
    exit 1
else
    echo "sha256 checksum matches"
fi
```

### I expected the following

I wish there was a`curl` flag, where curl would exit-non-zero, if an outputed file didn't match a specified hash.

```
curl --output-checksum sha256:XXXX -o /tmp/example.tar.gz http://.../
```

**Comment by jay** [comment]
Use --location and --fail so that it redirects on 3xx and fails on >= 4xx.

The only advantage I see to your feature request is the checksum could be calculated as the file is being downloaded instead of having to read it afterwards, which if the file is big could take some time. I think that's already possible if you build libcurl with metalink. Still just to see how this would work I am going to write a simpler draft for something like curl --write-out "%{sha256}" then a script could take that sha256 output and compare.

**Comment by bagder** [comment]
I'm not against the idea but I'm also questioning exactly how useful it would be in the end. Are sha256 hashes for download files really that commonly provided? Is this then going to just open the pandora's box for requests for other possible hash algorithms that are posted for some other downloadable packages? Isn't using `sha256sum` after the download is completed *almost* as easy as using a new curl option? A curl option that will take years until it is widely available anyway so scripts across the globe still need to use sha256sum for a long time into the future to make sure they can verify checksums...

**Comment by ghedo** [comment]
This is something I've also had to deal with in the past, but it's not hard to come up with something using standard command-line tools, so I don't quite feel the need for this to be supported by curl itself.

FTR here's what I came up with:

```
 % curl -fLs https://www.example.com | tee output.txt | sha256sum -c sum.txt
-: OK
 % echo $?
0
```

the resource will be downloaded into `output.txt` and `sum.txt` should contain a valid SHA256 checksum line for `-` (stdin), like:

```
 % cat sum.txt 
3587cb776ce0e4e8237f215800b7dffba0f25865cb84550e87ea8bbac838c423  -
```

Btw, on error it looks like this:
```
 % curl -fLs https://www.example.com | tee output.txt | sha256sum -c sum.txt 
-: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
 % echo $?
1
```

If you need to download multiple files you might be better off just calling `curl` a bunch of times and then calling `sha256sum` at the end once you downloaded all of them.

It's not quite the same as described in the original report, but hope it helps even if just a bit.

**Comment by jay** [comment]
All good points. Yes he could pipe the output and save time. I am intrigued by a more generic --match style option though.

I did end up writing a --write-out %{sha256} just to see how it would work, no match style. I'm not opening a PR for it or working on it further unless there's interest. It's largely untested, I did some basic tests in Visual Studio and that's it. It clocks in as 25 files changed, 520 insertions(+), 5 deletions(-) and uses the SHA-256 implementation found at https://github.com/ilvn/SHA256. Also, I put the heart of it into libcurl rather than just the tool, which may not be the best idea. See it at [master...jay:CURLINFO_SHA256](https://github.com/curl/curl/compare/master...jay:CURLINFO_SHA256?expand=1).

~~~
curld -V 

curl 7.54.0-DEV (i386-pc-win32) libcurl/7.54.0-DEV OpenSSL/1.0.2k nghttp2/1.21.0
Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS Debug TrackMemory Largefile NTLM SSL HTTP2 HTTPS-proxy SHA256

curld -sSfL --fail-early --write-out "%{sha256} *%{filename_effective}\n" --remote-name-all https://curl.haxx.se/download/curl-7.53.1.tar.gz https://curl.haxx.se/download/curl-7.53.1.tar.gz.asc

64f9b7ec82372edb8eaeded0a9cfa62334d8f98abc65487da01188259392911d *curl-7.53.1.tar.gz
5dff94f728493357bc71fcdcfce474bc64cf8c8e48d58175a39d6f767f38a808 *curl-7.53.1.tar.gz.asc
~~~
